# Advanced Programming Final Examination

In [2]:
# Student name: Ciniso Msibi
# Registrstion Number: MAY22/COMP/1423X
# Semester 1

#### a) The variable release_year_dict is a Python Dictionary, what is the result of applying the following method: release_year_dict.keys() ? (3 marks)

In [7]:
release_year_dict = {'Aa':'Advanced','Ab':'Programming'}
release_year_dict.keys()

dict_keys(['Aa', 'Ab'])

- It will retrieve the keys of the dictionary

#### b) In Python 3, what is the type of the variable x after the following: x=1/1 (3 marks)

In [4]:
x=1/1
x

1.0

- Float datatype

#### c) Consider the dataframe df how would you access the element in the 1st row 3rd column? (3 marks)

In [45]:
import pandas as pd

data = {"calories": [420, 380, 390,42],"duration": [50, 40, 45,45],"Weight": [70, 50, 95,41]}
df = pd.DataFrame(data)
df = df.iloc[[0],[2]]
df

Weight
0      70

- df.iloc[[o],[2]]

#### d) In the class, you learned you can also obtain a series from a dataframe df, write the correct way to assign the column with the header Length to a pandas series to the variable x. (3 marks)

- df.rename(columns = {"Length":"x"})

#### e) What is the result of applying the following method df.head(), to the dataframe df (3 marks)

- It shows the first 5 rows of the dataframe

#### f) How do you cast the list A to the set a? (3 marks)

- set_a = set(list_A)

#### g) What is the result of the following: 'C' in {'A','B'}? (3 marks)

- {'A','B','C'}

#### h) Consider the Set: V={'1','2'}, what is the result of V.add('3')? (3 marks)

In [48]:
V = {'1','2'}
V.add('3')
V

{'1', '2', '3'}

- {'1', '2', '3'}

#### i)  Consider the Set: U=['1','2'], write a code for adding 3 in set U? (3 marks)

- U.add('3')

#### j) What is the result of the following: str(1)+str(6)? (3 marks)

In [50]:
str(1)+str(6)

'16'

- '16'

#### k) Consider the variable B ={“WHO Region”, “Geographical location”, “longitude and latitude” }. Write python code that can change WHO Region to “Region”, Geographical location to “Location”, longitude and latitude to “Location”. (4 marks)

In [57]:
B = {"WHO Region", "Geographical location", "longitude and latitude" }
B.update(["Region"])
B

{'Geographical location', 'Region', 'WHO Region', 'longitude and latitude'}

#### l) Consider the string A='1293456789', what is the result of the following operation A[4::] (3 marks)


In [54]:
A ='1293456789'
A[4::]

'456789'

- '456789'

#### m) Consider the string Name="UTAMU", what is the result of the following operation Name.find("M"). (3 marks)

In [59]:
Name ="UTAMU"
Name.find("M")

3

- The result is the position of the letter M in the variable Name's value.The position is 3

## Section B

### Question 2

In [60]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium
print('Folium installed and imported!')

Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... 
  - anaconda/win-64::certifi-2021.10.8-py39haa95532_2, anaconda/win-64::openssl-1.1.1n-h2bbff1b_0
  - anaconda/win-64::openssl-1.1.1n-h2bbff1b_0, defaults/win-64::certifi-2021.10.8-py39haa95532_2
  - anaconda/win-64::certifi-2021.10.8-py39haa95532_2, defaults/win-64::openssl-1.1.1n-h2bbff1b_0
  - defaults/win-64::certifi-2021.10.8-py39haa95532_2, defaults/win-64::openssl-1.1.1n-h2bbff1b_0done

## Package Plan ##

  environment location: C:\Users\OXIDE_MUSICA\anaconda3

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    altair-4.2.0               |     pyhd8ed1ab_1         711 KB  conda-forge
    branca-0.5.0               |     pyhd8ed1ab_0          26 KB  conda-forge
    conda-build-3.22.0         |   py39hcbf53

In [72]:
Ugandan_map = folium.Map(location=[1.3733,32.2903], zoom_start=4)

#### i) Kampala

In [67]:
Kampala = folium.map.FeatureGroup()
Kampala.add_child(folium.features.CircleMarker([0.3476,32.5825], radius=5,color="red",fill_color="red"))
Ugandan_map.add_child(Kampala)
folium.Marker([0.3476,32.5825],popup="Kampala").add_to(Ugandan_map)

#### ii) Jinja

In [68]:
jinja = folium.map.FeatureGroup()
jinja.add_child(folium.features.CircleMarker([0.4479,33.2026], radius=5,color="red",fill_color="red"))
Ugandan_map.add_child(jinja)
folium.Marker([0.4479,33.2026],popup="Jinja").add_to(Ugandan_map)

#### iii) Gulu

In [69]:
gulu = folium.map.FeatureGroup()
gulu.add_child(folium.features.CircleMarker([2.7724,32.28816], radius=5,color="red",fill_color="red"))
Ugandan_map.add_child(gulu)
folium.Marker([2.7724,32.2881],popup="Gulu").add_to(Ugandan_map)

#### iv) Mbarara

In [70]:
mbarara = folium.map.FeatureGroup()
mbarara.add_child(folium.features.CircleMarker([0.6072,30.6545], radius=5,color="red",fill_color="red"))
Ugandan_map.add_child(mbarara)
folium.Marker([0.6072,30.6545],popup="Mbarara").add_to(Ugandan_map)

In [71]:
Ugandan_map

#### Question 4

In [73]:
def get_concat_h(im1, im2):
    #https://note.nkmk.me/en/python-pillow-concat-images/
    dst = Image.new('RGB', (im1.width + im2.width, im1.height))
    dst.paste(im1, (0, 0))
    dst.paste(im2, (im1.width, 0))
    return dst

In [74]:
my_image = "exam.jpeg"

In [75]:
import os
cwd = os.getcwd()
cwd

'C:\\Users\\OXIDE_MUSICA'

In [76]:
image_path = os.path.join(cwd, my_image)
image_path

'C:\\Users\\OXIDE_MUSICA\\exam.jpeg'

In [ ]:
import cv2

In [ ]:
image = cv2.imread(my_image)

In [ ]:
type(image)

In [ ]:
image.shape

In [ ]:
image.max()

In [ ]:
image.min()

In [ ]:
image_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

In [ ]:
image_gray.shape

In [ ]:
plt.figure(figsize=(10, 10))
plt.imshow(image_gray, cmap='gray')
plt.show()

## Question 5